assemblies/assembly-planner-configuration.adoc

```diff
@@ -0,0 +1,1986 @@
+ifdef::OPTAPLANNER-COMM[]
+image::PlannerConfiguration/solvedNQueens04.png[align="left"]
+endif::OPTAPLANNER-COMM[]
+ifdef::OPTAPLANNER-ENT[]
+image::optimizer/solvedNQueens04.png[align="left"]
+endif::OPTAPLANNER-ENT[]
+
+
+The `solve(Solution)` method can take a long time depending on the problem size and the solver configuration. The `Solver` intelligently works through the search space of possible solutions and remembers the best solution it encounters during solving.
+Depending on a number of factors, including problem size, how much time the `Solver` has, the solver configuration, and so forth, that best solution might or might not be an optimal solution.
+
+[NOTE]
+====
+The solution instance given to the method `solve(solution)` is changed by the ``Solver``, but it might not be the best solution.
+
+The solution instance returned by the methods `solve(solution)` or `getBestSolution()` is most likely a planning clone of the instance given to the method ``solve(solution)``, which implies it is a different instance.
+====
+
+
+
+[[environmentMode]]
+=== Solver environment mode
+
+The environment mode allows you to detect common bugs in your implementation.
+It does not affect the logging level.
+
+A solver has a single `Random` instance. Some solver configurations use the `Random` instance a lot more than others.
+For example, the Simulated Annealing algorithm depends highly on random numbers, while tabu search only depends on it to resolve score ties.
+The environment mode influences the seed of that `Random` instance.
+
+
+You can set the environment mode in the solver configuration XML file. The following example sets the `FAST_ASSERT` mode:
+
+[source,xml,options="nowrap"]
+----
+<solver xmlns="https://www.optaplanner.org/xsd/solver" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+    xsi:schemaLocation="https://www.optaplanner.org/xsd/solver https://www.optaplanner.org/xsd/solver/solver.xsd">
+  <environmentMode>FAST_ASSERT</environmentMode>
+  ...
+</solver>
+----
+
+
+.Environment modes:
+
+[[environmentModeFullAssert]]
+* `FULL_ASSERT` mode turns on all assertions (such as assert that the incremental score calculation is uncorrupted for each move) to fail-fast on a bug in a Move implementation, a constraint, the engine itself, and so on.  This mode is reproducible. It is also intrusive because it calls the method `calculateScore()` more frequently than a non-assert mode. The `FULL_ASSERT` mode is very slow because it does not rely on incremental score calculation.
+
+[[environmentModeNonIntrusiveFullAssert]]
+* `NON_INTRUSIVE_FULL_ASSERT` mode turns on several assertions to fail-fast on a bug in a Move implementation, a constraint, the engine itself, and so on. This mode is reproducible. It is non-intrusive because it does not call the method `calculateScore()` more frequently than a non assert mode.The `NON_INTRUSIVE_FULL_ASSERT` mode is very slow because it does not rely on incremental score calculation.
```

Comment:
```suggestion
* `NON_INTRUSIVE_FULL_ASSERT` mode turns on several assertions to fail-fast on a bug in a Move implementation, a constraint, the engine itself, and so on. This mode is reproducible. It is non-intrusive because it does not call the method `calculateScore()` more frequently than a non-assert mode.The `NON_INTRUSIVE_FULL_ASSERT` mode is very slow because it does not rely on incremental score calculation.
```